Models/IonAuthModel.php

```diff
@@ -0,0 +1,2798 @@
+				$this->setMessage('IonAuth.login_successful');
+
+				return true;
+			}
+		}
+
+		// Hash something anyway, just to take up time
+		$this->hashPassword($password);
+
+		$this->increaseLoginAttempts($identity);
+
+		$this->triggerEvents('post_login_unsuccessful');
+		$this->setError('IonAuth.login_unsuccessful');
+
+		return false;
+	}
+
+	/**
+	 * Verifies if the session should be rechecked according to the configuration item recheckTimer. If it does, then
+	 * it will check if the user is still active
+	 * @return bool
+	 */
+	public function recheckSession()
+	{
+		$recheck = (NULL !== $this->config->recheckTimer) ? $this->config->recheckTimer : 0;
+
+		if ($recheck !== 0)
+		{
+			$last_login = $this->session->get('last_check');
+			if ($last_login + $recheck < time())
+			{
+				$query = $this->db->select('id')
+								  ->where([
+									  $this->identity_column => $this->session->get('identity'),
+									  'active' => '1'
+								  ])
+								  ->limit(1)
+								  ->orderBy('id', 'desc')
+								  ->get($this->tables['users']);
+				if ($query->numRows() === 1)
+				{
+					$this->session->set_userdata('last_check', time());
+				}
+				else
+				{
+					$this->triggerEvents('logout');
+
+					$identity = $this->config->identity;
+
+					$this->session->unset_userdata([$identity, 'id', 'user_id']);
+
+					return false;
+				}
+			}
+		}
+
+		return (bool)session('identity');
+	}
+
+	/**
+	 * Check if max login attempts exceeded
+	 * Based on code from Tank Auth, by Ilya Konyukhov (https://github.com/ilkon/Tank-Auth)
+	 *
+	 * @param string      $identity  user's identity
+	 * @param string|null $ipAddress IP address
+	 *                               Only used if trackLoginIpAddress is set to TRUE.
+	 *                               If NULL (default value), the current IP address is used.
+	 *                               Use getLastAttemptIp($identity) to retrieve a user's last IP
+	 *
+	 * @return boolean
+	 */
+	public function isMaxLoginAttemptsExceeded(string $identity, string $ipAddress = null): bool
+	{
+		if ($this->config->trackLoginAttempts)
+		{
+			$maxAttempts = $this->config->maximumLoginAttempts;
+			if ($maxAttempts > 0)
+			{
+				$attempts = $this->getAttemptsNum($identity, $ipAddress);
+				return $attempts >= $maxAttempts;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Get number of login attempts for the given IP-address or identity
+	 * Based on code from Tank Auth, by Ilya Konyukhov (https://github.com/ilkon/Tank-Auth)
+	 *
+	 * @param string      $identity   User's identity
+	 * @param string|null $ip_address IP address
+	 *                                Only used if trackLoginIpAddress is set to TRUE.
+	 *                                If NULL (default value), the current IP address is used.
+	 *                                Use getLastAttemptIp($identity) to retrieve a user's last IP
+	 *
+	 * @return int
+	 */
+	public function getAttemptsNum(string $identity, string $ipAddress = null): int
```

Comment:
Remove typehint due to mixed
```suggestion
	public function getAttemptsNum(string $identity, $ipAddress = null): int
```